### DISCUSSION TITLE
enum members added by `enum.global_enum` to module space is not detected by pyright.

### DISCUSSION THREAD

**Discussion by Achxy** [microsoft/pyright#4239] [Q&A]
If enum members were exported to module global namespace by [`enum.global_enum`](https://docs.python.org/3/library/enum.html#enum.global_enum) they are not reflected by pyright and are considered as not defined.
Example:
```py
from enum import Enum, auto, global_enum


@global_enum
class Foo(Enum):
    BAR = auto()
    QUX = auto()


print(QUX)  # "QUX" is not defined
# However it is defined during runtime
````

**Comment by erictraut** [comment]
Looks like `enum.global_enum` was added in Python 3.11. This feature breaks the basic rules of encapsulation and allows a class to modify the global (module) namespace. I'm quite surprised that this feature was added.

I'm not inclined to support this feature in static type checking. It's not clear to me whether we could even support such a feature given the way that pyright is designed. As I mentioned, `global_enum` breaks a core assumption.

If, for whatever reason, you want to expose symbols in your global namespace and want this to be compatible with static type checkers, you should do so explicitly.

```python
class Foo(Enum):
    BAR = auto()
    QUX = auto()

BAR = Foo.BAR
QUX = Foo.QUX
```

**Reply by Achxy** [reply]
Since the values of an enum class cannot be patched once defined and that enum classes cannot be further subclassed should make this feature at least possible.
```py
class Foo(Enum):
    A = 1

Foo.A.value = 2 # Fails during runtime
```
---
```py
class Foo(Enum):
    A = 1

class Bar(Foo): # Fails during typecheck and during runtime
    B = 2
```
There are instances of this feature [being used in the standard library - cpython](https://github.com/python/cpython/blob/75a6fadf369315b27e12f670e6295cf2c2cf7d7e/Lib/re/__init__.py#L141-L156)
and implementation of this feature could aid in eloquence by [preventing such annotations - typeshed](https://github.com/python/typeshed/blob/ea7bbbdad7abf05a3bd3feb25eef93b53165e2d9/stdlib/re.pyi#L165-L202)
```py
A = RegexFlag.A
ASCII = RegexFlag.ASCII
DEBUG = RegexFlag.DEBUG
I = RegexFlag.I
IGNORECASE = RegexFlag.IGNORECASE
L = RegexFlag.L
LOCALE = RegexFlag.LOCALE
M = RegexFlag.M
MULTILINE = RegexFlag.MULTILINE
S = RegexFlag.S
DOTALL = RegexFlag.DOTALL
X = RegexFlag.X
VERBOSE = RegexFlag.VERBOSE
U = RegexFlag.U
UNICODE = RegexFlag.UNICODE
T = RegexFlag.T
TEMPLATE = RegexFlag.TEMPLATE
```
Making enum contents public may not adhere to encapsulation principles most of the time but it has some use-cases like observed in `re` and some popular libraries. So this may be a great feature to be implemented to support such cases.

**Reply by plokmijnuhby** [reply]
I totally agree with this. I suppose you'd need to special case enums to deal with it, similarly to how namedtuple is special cased, but I think that's worth doing.
